Adjust the storage validator to the new storage

ljedrz opened the pull request:
This PR re-enables the storage validator tool after the recent storage revamp.

Even though it appears to work, the PR is still a draft, as I'm still not sure if I used the best APIs, and due to the fact the validator-related tests are still disabled (as they are incompatible with the validator now).

My intention is for the validator to be as low-level as possible and to remain snappy ~(hence the `KeyValueStorage` API tweaks)~.

Protryon commented on storage/src/key_value/mod.rs:
```diff
@@ -33,18 +33,20 @@ use agent::Agent;
-    fn get_column_keys<'a>(&'a mut self, column: KeyValueColumn) -> Result<Vec<Value<'a>>>;
+    fn get_column_keys<'a>(&'a self, column: KeyValueColumn) -> Result<Vec<Value<'a>>>;
 
-    fn get_column<'a>(&'a mut self, column: KeyValueColumn) -> Result<Vec<(Value<'a>, Value<'a>)>>;
```
these must be `mut` to ensure proper thread safety. please do not change the mutability of these APIs. rocksdb crate doesn't ensure thread safety. Feel free to call rocksDB directly if you need it.

Protryon reviewed (changes requested):
We can't change the KeyValueStorage to non-mut members, it violates thread safety internal to RocksDB and changes the entire paradigm of what it's supposed to mean (single-threaded, strongly consistent). Use rocksDB directly if you want to call it from multiple threads.

RocksDB colloquially works if you read from several threads (while not writing) -- but we don't currently have an interface to support this. An option might be to request a read-only interface, taking the mutable reference, and pass this into a `Box<Fn>` or extract it somehow else.

Protryon commented:
LGTM -- in general, if we see any issues in storage (barring unmigrated/unmitigated issues from pre-refactor) it should be critically errored/reported as a bug.

ljedrz commented:
Ok, the PR is now in a shape good enough for merging - the most important validator-specific test and all the consensus sidechain tests now work and pass too.

Changes since the last review:
- make the storage validator apply to all `T: KeyValueStorage`
- change a duplicate lookup for `cm` to `digest` (a copy-paste error during the rewrite)
- change one instance of `Arc<dyn Storage>` to `DynStorage`
- enable storage validation for `DynStorage`
- exclude ledger digest indices in one of the storage validator checks (I think the indices hadn't been stored as values before)
- **store the pre-genesis digest** (this was also a bug before the storage refactor, fixed in https://github.com/AleoHQ/snarkOS/pull/884/commits/957338c4586ebbdff124b66bc3f7d6e57c994ba5)
- restore the storage validator tests (excluding the validator fix tests for now)

niklaslong reviewed (commented):
Works nicely for me locally (and with a fresh db) 👍

Protryon commented on storage/src/key_value/agent/mod.rs:
```diff
@@ -41,18 +42,23 @@ use crate::{
 
 pub(super) struct Agent<S: KeyValueStorage + 'static> {
-    inner: S,
+    inner: Option<S>, // the option is only for the purposes of validation, which requires ownership
```
can we just decompose the agent? an agent with no inner just panics anyways -- and this makes the code less clean.

Protryon commented on storage/src/key_value/agent/mod.rs:
```diff
@@ -142,13 +148,33 @@ impl<S: KeyValueStorage + 'static> Agent<S> {
 
+    fn validate(&mut self, limit: Option<u32>, fix_mode: FixMode) -> bool {
+        let result = if let Some(inner) = mem::take(&mut self.inner) {
+            let (ret, inner) = futures::executor::block_on(inner.validate(limit, fix_mode));
```
please be super, super careful with block_on, and try to avoid it if possible. it can create deadlocks very easily.

Protryon replied:
i question the validator needing ownership (why not a mut ref?) -- also if it does need ownership, clone or even better, just take self on `validate` calls since we terminate execution after validating anyways.

Protryon replied:
also why does it need to be async if presumably the only place calling it is sync?

ljedrz replied:
this function is intended to only ever be called in some tests (and those have been working well), so we should be fine; good point, of course

Protryon replied:
what do you think about taking ownership then? i.e. `self` target

ljedrz replied:
>also why does it need to be async if presumably the only place calling it is sync?
I guess I could use `try_send` in the associated method (which should remove the need for `async`), let me check :+1:

ljedrz replied:
>what do you think about taking ownership then? i.e. self target

I could try it out

ljedrz replied:
as discussed externally, these changes were currently not viable; they might be possible with planned future changes to the storage

Protryon commented on storage/src/storage.rs:
```diff
@@ -160,6 +163,9 @@ pub trait Storage: Send + Sync {
     async fn get_block_hashes(&self, limit: Option<u32>, filter: BlockFilter) -> Result<Vec<Digest>>;
+
+    /// Performs low-level storage validation; it's mostly intended for test purposes, as there is a lower level `KeyValueStorage` interface available outside of them.
+    async fn validate(&self, limit: Option<u32>, fix_mode: FixMode) -> bool;
```
we should make the docs here generic as to future implementations, which would presumably validate themselves as well -- maybe in a strongly typed relational DB that would be a nop, but thats okay.
